README.md

```diff
@@ -59,12 +69,71 @@ The FetchMigration directory hosts tools that simplify the process of backfillin
 
 Further documentation can be found here: [Fetch Migration README](FetchMigration/README.md).
 
-### Running Tests
+### Testing 
 
-Developers can run a test script which will verify the end-to-end Local Docker Solution.
+Migration assisant is built for being run in a cloud environment to scale up to large clusters of all kinds.  This creates challenges to quickly validate and detect failures introduced by changes.  This project balances these scenarios with two forms of 'end to end' testing, cloud and local end to end testing.
 
-More documentation on this test script can be found here:
-[End-to-End Testing](test/README.md)
+#### Component and Unit testing
+
+While end to end tests can be very useful for testing integration boundaries, for simplier components and functions it is advised to write junit tests that focus on edge cases.  When adding functionality its advised to start at these lower level test and then see what end to end tests might need to be included or modified to check for coverage if needed.
+
+#### Cloud end to end testing
+
+These tests are setup once with a deployment into a cloud ecosystem, because full infrastructure is inheriantly expensive monitarily and time wise - there will be only a small number of tests that make sure the migrations tools are working as expected.  In this example diagram, tests to validate historical migration from a snapshot use the full cloud resources.  Control of the test scenarios is largely through the deployed migration console.
+
+See [End-to-End Testing](test/README.md) for more details on deployment and execution of these test cases.
+
+```mermaid
+graph LR;
+    Test[Python Tests] --> MC
+    subgraph AWSCloud
+        MC[Migration Console]
+        SC[Source Cluster]
+        TC[Target Cluster]
+        S3[S3 Snapshot]
+
+        subgraph ECS
+            HW[Historical Workers - RFS]
+        end
+
+        HW --> SC
+        HW --> TC
+        HW --> S3
+
+        MC --> ECS
+        MC --> SC
+        MC --> TC
+    end
+```
+
+#### Local end to end testing
+
+Locally performing end to end test cases is done slightly differently, by subtituting cloud services for the local file system, and TestContainers to provider source and target migration clusters its much faster to validate different configurations and versions.  By running inside of a JUnit JVM this also allows deep access into the running historical worker for monitoring or fault injection.  When test cases themselves do not require full source or target clusters using mock endpoints or data can greatly accelerate test speed.
+
+These tests are all be run with the command `./gradlew test` from the root directory.
```

Comment:
Hmm, what about some of the python code we have, seems like we are using the python test runner - does that apply here or is do we think that is basically covered in the component/unit test kind of area.

Reply:
Currently, we use python tests both for unit testing (e.g. console lib or fetch), as well as to run our existing end to end tests. It might be possible to run the latter from within gradle (though I'm not sure we gain much from that), but my understanding from this diagram (and the emphasis on running the tests in the JVM) is that you're suggesting rewriting/porting/scrapping those tests. Is that accurate? If so, running python tests might not be that much of a concern.

Is that what you're asking here, or are you wondering if there's a role that we're using python for that java tests couldn't fill?

Reply:
Good distinction, python is well suited for the defining the console link CLI and API.  This code is hard to test beyond the unit level with anything other than the real control plane systems (be it a cloud service, kuberneties, or local docker) and java-based migration tools.  What do you think of this?

I guess I'm saying that testing the python code in an end to end way involves `Cloud end to end testing` kind of structure rather than another category - I am curious we think there is a way to test the python codebase that is missing here. (Fake java migration tools that spit out errors?)

Reply:
Yep, I'm fine with that. I think we definitely could (should?) do some testing of the CLI and API that involves fake/toy containers running locally , but I don't think it's a particularly high priority to make that a local thing instead of a cloud-level thing.